sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java

```diff
@@ -21,38 +21,232 @@
+ *       based on runner behavior, though at least 1 shard will always be produced for every
+ *       non-empty window. Note that setting a fixed number of shards can hurt performance: it adds
+ *       an additional {@link GroupByKey} to the pipeline.
+ *   <li><b>How the shards are named:</b> This is controlled by a {@link Write.FilenamePolicy}:
+ *       filenames can depend on a variety of inputs, e.g. the window, total number of shards and
+ *       the current file's shard index, etc. - see {@link Write.FilenameContext}. The standard
+ *       policies are {@link Write#prefixAndWindowedShard} and {@link
+ *       Write#prefixAndUnwindowedShard}. A concise way to create a custom policy is {@link
+ *       Write#prefixAndShardTemplate}.
+ *   <li><b>Which elements go into which shard:</b> It is not possible to control this - elements
+ *       within a window get distributed into different shards created for that window arbitrarily,
+ *       though {@link FileIO.Write} attempts to make shards approximately evenly sized. For more
+ *       control over which elements go into which files, consider using <i>dynamic destinations</i>
+ *       (see below).
+ *   <li><b>How a given set of elements is written to a shard:</b> This is controlled by the {@link
+ *       Sink}, e.g. {@link AvroIO#sink} will generate Avro files. The {@link Sink} controls the
+ *       format of a single file: how to open a file, how to write each element to it, and how to
+ *       close the file - but it does not control the set of files or which elements go where.
+ *       Elements are written to a shard in an arbitrary order. {@link FileIO.Write} can
+ *       additionally compress the generated files using {@link FileIO.Write#withCompression}.
+ *   <li><b>How all of the above can be element-dependent:</b> This is controlled by <i>dynamic
+ *       destinations</i>. It is possible to have different groups of elements use different
+ *       policies for naming files and for configuring the {@link Sink}. See "dynamic destinations"
+ *       below.
+ * </ul>
+ *
+ * <h3>Dynamic destinations</h3>
+ *
+ * <p>If the elements in the input collection can be partitioned into groups that should be treated
+ * differently, {@link FileIO.Write} supports different treatment per group ("destination"). It can
+ * use different filename policies for different groups, and can differently configure the {@link
+ * Sink}, e.g. write different elements to Avro files in different directories with different
+ * schemas.
+ *
+ * <p>This feature is supported by {@link #writeDynamic}. Use {@link Write#by} to specify how to
+ * partition the elements into groups ("destinations"). Then elements will be grouped by
+ * destination, and {@link Write#to(SerializableFunction)} and the {@link
+ * Write#via(SerializableFunction)} will be applied separately within each group. Note that
+ * currently sharding can not be destination-dependent.
```

Comment:
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
Please explain this.